modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+    private IByteArrayKeyValueDatabase indexSource;
+
+    // tracking the status
+    private Map<ByteArrayWrapper, QueueInfo> status;
+    private long maxRequest = 0L, minStatus = Long.MAX_VALUE, maxStatus = 0L;
+
+    private static final int FORWARD_SKIP = STEP_COUNT * LARGE_REQUEST_SIZE;
+
+    /**
+     * Constructor. Initializes the databases used for storage. If the database configuration used
+     * requires persistence, the constructor ensures the path can be accessed or throws an exception
+     * if persistence is requested but not achievable.
+     *
+     * @param _props properties of the databases to be used for storage
+     * @throws InvalidFilePathException when given a persistent database vendor for which the path
+     *     cannot be created
+     */
+    public PendingBlockStore(final Properties _props) throws InvalidFilePathException {
+        Properties local = new Properties(_props);
+
+        // check for database persistence requirements
+        DBVendor vendor = DBVendor.fromString(local.getProperty(Props.DB_TYPE));
+        if (vendor.getPersistence()) {
+            File pbFolder =
+                    new File(local.getProperty(Props.DB_PATH), local.getProperty(Props.DB_NAME));
+
+            verifyAndBuildPath(pbFolder);
+            local.setProperty(Props.DB_PATH, pbFolder.getAbsolutePath());
+        }
+
+        init(local);
+    }
+
+    /**
+     * Initializes and opens the databases where the pending blocks will be stored.
+     *
+     * @param props the database properties to be used in initializing the underlying databases
+     */
+    private void init(Properties props) {
+        // initialize status
+        this.status = new HashMap<>();
+
+        // create the level source
+        props.setProperty(Props.DB_NAME, LEVEL_DB_NAME);
+        this.levelDatabase = connectAndOpen(props, LOG);
+        this.levelSource = new ObjectDataSource<>(levelDatabase, HASH_LIST_RLP_SERIALIZER);
+
+        // create the queue source
+        props.setProperty(Props.DB_NAME, QUEUE_DB_NAME);
+        this.queueDatabase = connectAndOpen(props, LOG);
+        this.queueSource = new ObjectDataSource<>(queueDatabase, BLOCK_LIST_RLP_SERIALIZER);
+
+        // create the index source
+        props.setProperty(Props.DB_NAME, INDEX_DB_NAME);
+        this.indexSource = connectAndOpen(props, LOG);
+    }
+
+    /**
+     * Checks that the underlying storage was correctly initialized and open.
+     *
+     * @return true if correctly initialized and the databases are open, false otherwise.
+     */
+    public boolean isOpen() {
+        return status != null
+                && levelSource.isOpen()
+                && queueSource.isOpen()
+                && indexSource.isOpen();
+    }
+
+    private static final Serializer<List<byte[]>, byte[]> HASH_LIST_RLP_SERIALIZER =
+            new Serializer<>() {
+                @Override
+                public byte[] serialize(List<byte[]> object) {
```

Comment:
I think it's good the including the jsr305 library to check the nonnull status (adding @Nonnull in front of the inputting args )

Reply:
for the Serializer class, do we need to confirm the input data has no null object?

Reply:
`RLP.encodeElement` and `RLP.encodeList` will produce an encoding even if the input data is null. Also, both serializers are private to the class which makes sure to give them correct data, so I would prefer not to add the overhead of the annotations.

Reply:
I see